lib/msal-node/src/cache/CacheManager.ts

```diff
@@ -0,0 +1,229 @@
+     * Merge in memory cache with the cache snapshot.
+     * @param oldState
+     * @param currentState
+     */
+    private mergeState(oldState: JsonCache, currentState: JsonCache): JsonCache {
+        let stateAfterRemoval = this.mergeRemovals(oldState, currentState);
+        return this.mergeUpdates(stateAfterRemoval, currentState);
+    }
+
+    /**
+     * Deep update of oldState based on newState values
+     * @param oldState
+     * @param newState
+     */
+    private mergeUpdates(oldState: any, newState: any): JsonCache {
+
+        let finalState = oldState;
+        Object.keys(newState).forEach((newKey) => {
+            let newValue = newState[newKey];
+
+            // if oldState does not contain value but newValue does, add it
+            if (!finalState.hasOwnProperty(newKey)) {
+                if (newValue != null) {
+                    finalState[newKey] = newValue;
+                }
+            } else {
+                // both oldState and newState contain the key, do deep update
+                let newValueNotNull = newValue !== null;
+                let newValueIsObject = typeof newValue === 'object';
+                let newValueIsNotArray = !Array.isArray(newValue);
+
+                if (newValueNotNull && newValueIsObject && newValueIsNotArray) {
+                    this.mergeUpdates(finalState[newKey], newValue);
+                } else {
+                    finalState[newKey] = newValue;
+                }
+            }
+        });
+
+        return finalState;
+    }
+
+    /**
+     * Removes entities in oldState that the were removed from newState. If there are any unknown values in root of
+     * oldState that are not recognized, they are left untouched.
+     * @param oldState
+     * @param newState
+     */
+    private mergeRemovals(oldState: JsonCache, newState: JsonCache): JsonCache {
+        const accounts = oldState.Account != null ? this.mergeRemovalsStringDict(oldState.Account, newState.Account) : oldState.Account;
+        const accessTokens = oldState.AccessToken != null ? this.mergeRemovalsStringDict(oldState.AccessToken, newState.AccessToken) : oldState.AccessToken;
+        const refreshTokens = oldState.RefreshToken != null ? this.mergeRemovalsStringDict(oldState.RefreshToken, newState.RefreshToken) : oldState.RefreshToken;
+        const idTokens = oldState.IdToken != null ? this.mergeRemovalsStringDict(oldState.IdToken, newState.IdToken) : oldState.IdToken;
+        const appMetadata = oldState.AppMetadata != null ? this.mergeRemovalsStringDict(oldState.AppMetadata, newState.AppMetadata) : oldState.AppMetadata;
+
+        return {
+            Account: accounts,
+            AccessToken: accessTokens,
+            RefreshToken: refreshTokens,
+            IdToken: idTokens,
+            AppMetadata: appMetadata
+        }
+    }
+
+    private mergeRemovalsStringDict(oldAccounts: StringDict, newAccounts?: StringDict): StringDict {
+        let finalAccounts = oldAccounts;
```

Comment:
Same comment here about creating a copy so that you don't modify the passed in value.

Reply:
Updated